ruleset.go

```diff
@@ -24,15 +28,151 @@ const (
+	squads map[string]string // Snake ID -> Squad Name
+}
+
+// NewRulesetBuilder returns an instance of a builder for the Ruleset types.
+func NewRulesetBuilder() *rulesetBuilder {
+	return &rulesetBuilder{
+		params: map[string]string{},
+		squads: map[string]string{},
+	}
+}
+
+// WithParams accepts a map of game parameters for customizing games.
+//
+// Parameters are copied. If called multiple times, parameters are merged such that:
+//   - existing keys in both maps get overwritten by the new ones
+//   - existing keys not present in the new map will be retained
+//   - non-existing keys only in the new map will be added
+//
+// Unrecognised parameters will be ignored and default values will be used.
+// Invalid parameters (i.e. a non-numerical value where one is expected), will be ignored
+// and default values will be used.
+func (rb *rulesetBuilder) WithParams(params map[string]string) *rulesetBuilder {
+	for k, v := range params {
+		rb.params[k] = v
+	}
+	return rb
+}
+
+// WithSeed sets the seed used for randomisation by certain game modes.
+func (rb *rulesetBuilder) WithSeed(seed int64) *rulesetBuilder {
+	rb.seed = seed
+	return rb
+}
+
+// AddSnakeToSquad adds the specified snake (by ID) to a squad with the given name.
+// This configuration may be ignored by game modes if they do not support squads.
+func (rb *rulesetBuilder) AddSnakeToSquad(snakeID, squadName string) *rulesetBuilder {
+	rb.squads[snakeID] = squadName
+	return rb
+}
+
+// Ruleset constructs a customised ruleset using the parameters passed to the builder.
+func (rb rulesetBuilder) Ruleset() Ruleset {
+	standardRuleset := &StandardRuleset{
+		FoodSpawnChance:     paramsInt32(rb.params, ParamFoodSpawnChance, 0),
+		MinimumFood:         paramsInt32(rb.params, ParamMinimumFood, 0),
+		HazardDamagePerTurn: paramsInt32(rb.params, ParamHazardDamagePerTurn, 0),
+	}
+
+	name, ok := rb.params[ParamGameType]
+	if !ok {
+		return standardRuleset
+	}
+
+	switch name {
+	case GameTypeConstrictor:
+		return &ConstrictorRuleset{
+			StandardRuleset: *standardRuleset,
+		}
+	case GameTypeRoyale:
+		return &RoyaleRuleset{
+			StandardRuleset:   *standardRuleset,
+			Seed:              rb.seed,
+			ShrinkEveryNTurns: paramsInt32(rb.params, ParamShrinkEveryNTurns, 0),
+		}
+	case GameTypeSolo:
+		return &SoloRuleset{
+			StandardRuleset: *standardRuleset,
+		}
+	case GameTypeWrapped:
+		return &WrappedRuleset{
+			StandardRuleset: *standardRuleset,
+		}
+	case GameTypeSquad:
+		squadMap := map[string]string{}
+		for id, squad := range rb.squads {
+			squadMap[id] = squad
+		}
+		return &SquadRuleset{
+			StandardRuleset:     *standardRuleset,
+			SquadMap:            squadMap,
+			AllowBodyCollisions: paramsBool(rb.params, ParamAllowBodyCollisions, true),
+			SharedElimination:   paramsBool(rb.params, ParamSharedElimination, true),
+			SharedHealth:        paramsBool(rb.params, ParamSharedHealth, true),
+			SharedLength:        paramsBool(rb.params, ParamSharedLength, true),
```

Comment:
I see what you're saying. The reason I did this was because the old code was defaulting these to true in most places. If these weren't set they'd all be true. I'm pretty sure I had some failing tests when I originally implemented with them default to `false` as well.

Also, I think this is safe because these values won't be applied unless the game type is specifically `squad`.